compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -219,46 +168,108 @@ object Semantic:
+     *  environment in the key, because the environment is always hot.
+     *
+     *  We do not need the heap in the key, because the value of an expression
+     *  is only determined by the value of `this`. The heap is immutable: the
+     *  abstract values for object fields never change within one iteration.
+     *  The initial abstraction of a field is always a safe over-approximation
+     *  thanks to monotonicity of initialization states.
+     *
+     *  If the heap is unstable in an iteration, the cache should also be
+     *  unstable. This is because all values stored in the heap are also present
+     *  in the cache. Therefore, we only need to track whether the cache is
+     *  stable between two iterations.
+     *
+     *  The heap is not part of the fixed point computation -- we throw the
+     *  unstable heap from last iteration away. In contrast, we use the unstable
+     *  output cache from the last iteration as input for the next iteration.
+     *  This is safe because the heap is determined by the cache -- it is a
+     *  "local" data to the computing function, conceptually. Local data is
+     *  always safe be discarded.
+     *
+     *  Now, if a fixed point is reached, the local data contains stable data
+     *  that could be reused to check other classes. We employ this trick to
+     *  improve performance of the analysis.
+     */
+    private type ExprValueCache = Map[Value, Map[TreeWrapper, Value]]
+
+    /** The heap for abstract objects
+     *
+     *  The heap objects are immutable.
+     */
     private type Heap = Map[Ref, Objekt]
 
+    /** A wrapper for trees for storage in maps based on referential equality of trees. */
+    private abstract class TreeWrapper:
+      def tree: Tree
+
+      override final def equals(other: Any): Boolean =
+        other match
+        case that: TreeWrapper => this.tree eq that.tree
+        case _ => false
+
+      override final def hashCode = tree.hashCode
+
+    /** The immutable wrapper is intended to be stored as key in the heap. */
+    private class ImmutableTreeWrapper(val tree: Tree) extends TreeWrapper
+
+    /** For queries on the heap, reuse the same wrapper to avoid unnecessary allocation. */
+    private class MutableTreeWrapper extends TreeWrapper:
+      var queryTree: Tree | Null = null
+      def tree: Tree = queryTree match
+        case tree: Tree => tree
+        case null => ???
+
     class Cache:
-      private var last: CacheStore =  mutable.Map.empty
-      private var current: CacheStore = mutable.Map.empty
-      private val stable: CacheStore = mutable.Map.empty
+      /** The cache for expression values from last iteration */
+      private var last: ExprValueCache =  Map.empty
+
+      /** The updated cache for expression values based on the cache values from the last iteration
+       *
+       *  Both `last` and `current` are required to make sure an expression is evaluated once in each iteration.
```

Comment:
This sentence is not clear: how do last/current ensure an expression is evaluated once? And should it be "each expression" instead of "an expression" (or are there particular expressions you have in mind)?

Reply:
Yes, more precisely it should be "each expression reachable from the constructor of the current class modulo local reasoning".